libminifi/src/utils/SystemCPUUsageTracker.cpp

```diff
@@ -0,0 +1,190 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "utils/SystemCPUUsageTracker.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace utils {
+#ifdef __linux__
+
+SystemCPUUsageTracker::SystemCPUUsageTracker() :
+    total_user_(0), previous_total_user_(0),
+    total_user_low_(0), previous_total_user_low_(0),
+    total_sys_(0), previous_total_sys_(0),
+    total_idle_(0), previous_total_idle_(0) {
+  queryCPUTimes();
+}
+
+double SystemCPUUsageTracker::getCPUUsageAndRestartCollection() {
+  queryCPUTimes();
+  if (isCurrentQuerySameAsPrevious() || isCurrentQueryOlderThanPrevious()) {
+    return -1.0;
+  } else {
+    return getCPUUsageBetweenLastTwoQueries();
+  }
+}
+
+void SystemCPUUsageTracker::queryCPUTimes() {
+  previous_total_user_ = total_user_;
+  previous_total_user_low_ = total_user_low_;
+  previous_total_sys_ = total_sys_;
+  previous_total_idle_ = total_idle_;
+  FILE* file = fopen("/proc/stat", "r");
+  if (fscanf(file, "cpu %lu %lu %lu %lu", &total_user_, &total_user_low_, &total_sys_, &total_idle_) != 4) {
+    total_user_ = previous_total_user_;
+    total_user_low_ = previous_total_user_low_;
+    total_idle_ = previous_total_idle_;
+    total_sys_ = previous_total_sys_;
+  }
+  fclose(file);
+}
+
+bool SystemCPUUsageTracker::isCurrentQueryOlderThanPrevious() const {
+  return (total_user_ < previous_total_user_ ||
+          total_user_low_ < previous_total_user_low_ ||
+          total_sys_ < previous_total_sys_ ||
+          total_idle_ < previous_total_idle_);
+}
+
+bool SystemCPUUsageTracker::isCurrentQuerySameAsPrevious() const {
+  return (total_user_ == previous_total_user_ &&
+          total_user_low_ == previous_total_user_low_ &&
+          total_sys_ == previous_total_sys_ &&
+          total_idle_ == previous_total_idle_);
+}
+
+double SystemCPUUsageTracker::getCPUUsageBetweenLastTwoQueries() const {
+  uint64_t total_user_diff = total_user_ - previous_total_user_;
+  uint64_t total_user_low_diff = total_user_low_ - previous_total_user_low_;
+  uint64_t total_system_diff = total_sys_ - previous_total_sys_;
+  uint64_t total_idle_diff = total_idle_ - previous_total_idle_;
+  uint64_t total_diff =  total_user_diff + total_user_low_diff + total_system_diff;
+  double percent = static_cast<double>(total_diff)/static_cast<double>(total_diff+total_idle_diff);
```

Comment:
That wont happened because this function (protected) is only called after the various checks (isCurrentQueryOlderThanPrevious and isCurrentQuerySameAsPrevious) which make sure that there are valid differences between the last and current query.